sycl/include/sycl/ext/intel/experimental/esimd/detail/test_proxy.hpp

```diff
@@ -0,0 +1,106 @@
+//
+// We are using non static data member initialization approach to force
+// the value required. Initialization will take place even if no
+// default/copy/move constructor of the test_proxy class was explcitly
+// called by any of the user-defined constructors of the proxy target
+#define __ESIMD_DECLARE_TEST_PROXY                                             \
+  esimd::detail::test_proxy M_testProxy = esimd::detail::test_proxy();
+
+// Declare the getter to access the proxy from the tests
+#define __ESIMD_DECLARE_TEST_PROXY_ACCESS                                      \
+  const auto &get_test_proxy() const { return M_testProxy; }
+
+// Test proxy will be handled in a proper way by default/implicit move
+// constructors and move operators.
+// Still the user-defined constructors or move operators should explicitly state
+// what to do with each of class atributes, so a proper wrapper required
+#define __esimd_move_test_proxy(other)                                         \
+  do {                                                                         \
+    M_testProxy = std::move(other.M_testProxy);                                \
+  } while (false);
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace ext {
+namespace intel {
+namespace experimental {
+namespace esimd {
+namespace detail {
+
+/// The test_proxy class.
+///
+/// This is a helper class for tests to differentiate between the copy
+/// constructor/assignment and the move constructor/assignment calls,
+/// as the copy constructor works as the default fallback for every case with
+/// move constructor disabled or not provided
+///
+/// The test proxy is enabled only if the __ESIMD_ENABLE_TEST_PROXY macro is
+/// defined.
+/// It is expected for the class with the test proxy (the class under test) to:
+/// - provide the get_test_proxy() method
+/// - properly handle moving the test_proxy member in user-defined move
+///   constructors and user-defined assignment operators
+///
+/// Therefore the following expression is expected to return `true` only if the
+/// move constructor or move operator was called for the instance of the class
+/// under test:
+///   instance.get_test_proxy().was_moved()
+///
+/// \ingroup sycl_esimd
+class test_proxy {
+  // Define the default value to use for every constructor
+  bool m_moved = false;
+
+public:
+  test_proxy() { __esimd_dbg_print(test_proxy()); }
+
+  test_proxy(const test_proxy &) {
+    __esimd_dbg_print(test_proxy(const test_proxy &other));
+  }
+  test_proxy(test_proxy &&) {
+    __esimd_dbg_print(test_proxy(test_proxy && other));
+    m_moved = true;
+  }
+  test_proxy &operator=(const test_proxy &) {
+    __esimd_dbg_print(test_proxy::operator=(const test_proxy &other));
+    return *this;
+  }
+  test_proxy &operator=(test_proxy &&) {
+    __esimd_dbg_print(test_proxy::operator=(test_proxy &&other));
+    m_moved = true;
+    return *this;
+  }
+  bool was_moved() const { return m_moved; }
```

Comment:
sounds somewhat misleading. `m_moved` does not indicate that `*this` was moved. It rather tells that `*this` was either
- constructed via a move constructor or
- another object was moved into it
`was_move_destination()`? (\+ `was_created_with_move_ctor()` if we need to distinguish the two)

Reply:
Nope, we don't need to distinguish the move constructor and move assignment. So something like `was_created_by_move()` might work, though I'm not sure it looks good in case of the move assignment.

Not sure we should do it though.
It's UB to access source instance after it has been moved to the destination instance, so in terms of logic we might say that in case of `std::move` it's just the same instance moving around.

It's even more clear if we took into account the corner cases with the chained calls, for example:
```
// Chained move
test_proxy a;
test_proxy b(std::move(a));
test_proxy c(std::move(b));
assert(a.was_moved()); 		// UB, logically it's moved to "c"
assert(c.was_moved()); 		// valid: that's instance "a" that moved here

// Chained move and copy
test_proxy a;
test_proxy b(std::move(a));
test_proxy c(b);
assert(a.was_moved()); 		// UB, logically it's moved to "b"
assert(b.was_moved()); 		// valid: that's instance "a" that moved here
assert(c.was_moved() == false);	// valid: that's new instance "c" with data from "a"
```

Does it address your concerns or is it better to change it to the `was_created_by_move()`?

Reply:
Renamed to the suggested `was_move_destination()` on the second thought by 0c83d18279ec8f3642413f914ede6bafb5db28bc. Also aligned a class attribute code style with other ESIMD API headers.